src/Models/IoTWorkspaceProject.ts

```diff
@@ -203,19 +206,15 @@ export class IoTWorkspaceProject extends IoTWorkbenchProjectBase {
   }
 
   /**
-   * Create and load device component according to board id.
-   * Push device to component list.
+   * Init device components according to board id.
+   * Update component list with device.
    * @param boardId board id
    * @param scaffoldType scaffold type
    * @param templateFilesInfo template files info to scaffold files for device
    */
-  private async initDevice(
-    boardId: string,
-    scaffoldType: ScaffoldType,
-    templateFilesInfo?: TemplateFileInfo[]
-  ): Promise<void> {
-    if (!(await FileUtility.directoryExists(scaffoldType, this.deviceRootPath))) {
-      throw new DirectoryNotFoundError(
+  private initDeviceComponents(boardId: string, templateFilesInfo?: TemplateFileInfo[]): void {
+    if (!this.deviceRootPath) {
+      throw new ArgumentEmptyOrNullError(

```

Comment:
This is a little bit weird here to check a dependent class member "deviceRootPath" and throw error if it is empty or null. I would suggest putting such required value in the function parameter. Or ensure the value of `deviceRootPath` in the class constructor and no more check its value in this private function.

Reply:
There is no strict principle on the problem. I would rather follow the below principles:

1. whether validate parameters inside functions:
* Document the methods well, and encourage the callers to pass good/valid data.
* the parameters anyway! - throwing exceptions when the preconditions are not met.(Unless the function is called too many times that checking parameters will cause huge time costs. E.g. Checking parameters when calculate matrix, etc.)

2. Whether passing private members as parameters into private methods:
* If the private parameter is closely logically relevant with the private method, it can serve as a parameter of the function.

> E.g. int area(int length, int width);
	
* If the private members is more of a global state, then don't pass it as parameter.

In our code, I prefer to put `deviceRootPath` as parameter since it is not initialized in the class constructor, but initialized differently from `create()` and `load()` methods. For other used private parameters, like `this.extensionContext`, `this.channel`, `this.telemetryContext`, I would rather keep them **NOT** as parameter but being accessed directly from the private function.